Android `background_color` Transparency Fixes

proneon267 opened the pull request:
<!--- Describe your changes in detail -->
Originally identified in #2484, this PR separates the changes required to fix the `background_color` fixes on Android.

* Closes #1620

_Originally posted in: https://github.com/beeware/toga/pull/2484#issuecomment-2137765659 :_
---
As I had noted here:
https://github.com/beeware/toga/blob/c9c714e3f05d48183a1ace86745812dd6103e3d2/android/src/toga_android/widgets/base.py#L197-L212 :

What I meant by that is that in some widgets, for example: `toga.Selection`, when their background_color is set via `set_background_simple` then the native effects are lost.
So, if I have:
```
toga.Box(
            style=Pack(flex=1),
            children=[
                toga.Selection(
                    items=["Alice", "Bob", "Charlie"],
                )
            ],
        )
```
Then the animation effect on interaction would look like:
![](https://github.com/beeware/toga/assets/45512885/a878e985-774f-48e6-be0d-bc6e07b4a911)
But now, if I set a background color on the parent like:
```
toga.Box(
            style=Pack(flex=1, background_color="#87CEFA"),
            children=[
                toga.Selection(
                    items=["Alice", "Bob", "Charlie"],
                )
            ],
        )
```
Then the animation effect on interaction would look like:
![](https://github.com/beeware/toga/assets/45512885/ecb6b749-712a-41ca-b8fb-06a569631cb9)
As we can see, the dropdown arrow is not visible, and the ripple effect is also not visible. Further, the widget doesn't set any background color and hence even setting the parent's background color also destroys the native effects. The same problem exists in other widgets also(like `toga.Switch`, etc.), where the ripple effect will not be visible:
![](https://github.com/beeware/toga/assets/45512885/c16fa158-26da-4de2-9811-66e3dc50bb33)
![](https://github.com/beeware/toga/assets/45512885/4aecf250-1f5e-4feb-9b63-18a779cea670)

As I have previously noted in the comment, tampering with the background(`LayerDrawable`) destroys it and with it the animation effects are also destroyed. Hence, I tried to solve this with `ContainedWidget` class which doesn't require the existing widgets to be changed.

There is also, `toga.DetailedList`:
```
        toga.Box(
            style=Pack(flex=1),
            children=[
                toga.DetailedList(
                    data=[
                        {
                            "icon": toga.Icon("icons/arthur"),
                            "title": "Arthur Dent",
                            "subtitle": "Where's the tea?",
                        },
                        {
                            "icon": toga.Icon("icons/ford"),
                            "title": "Ford Prefect",
                            "subtitle": "Do you know where my towel is?",
                        },
                        {
                            "icon": toga.Icon("icons/tricia"),
                            "title": "Tricia McMillan",
                            "subtitle": "What planet are you from?",
                        },
                    ],
                ),
            ],
        )
```
![](https://github.com/beeware/toga/assets/45512885/960aa3fc-6e27-449d-b96d-0a3081826efd)
But now, if I set a background color on the parent like:
```
        toga.Box(
            style=Pack(flex=1, background_color="#87CEFA"),
            children=[
                toga.DetailedList(...)
            ],
        )
```
![](https://github.com/beeware/toga/assets/45512885/b15f3d66-c090-4194-b342-ba881baf6d6c)
Similar, problems also exist on other widgets also(like `toga.Canvas`). Hence, for these widgets also, I have used`ContainedWidget` class to solve these issues.

<!--- What problem does this change solve? -->
<!--- If this PR relates to an issue, include Refs #XXX or Fixes #XXX -->

## PR Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

proneon267 commented:
> if I set a background color on the parent like:
> 
> ```
> toga.Box(
>             style=Pack(flex=1, background_color="#87CEFA"),
>             children=[
>                 toga.Selection(
>                     items=["Alice", "Bob", "Charlie"],
>                 )
>             ],
>         )
> ```
> 
> Then the animation effect on interaction would look like:
> ![](https://github.com/beeware/toga/assets/45512885/ecb6b749-712a-41ca-b8fb-06a569631cb9)
> 
> As we can see, the dropdown arrow is not visible, and the ripple effect is also not visible. Further, the widget doesn't set any background color and hence even setting the parent's background color also destroys the native effects.

I can confirm that this also happens Toga 0.4.5, so it wasn't introduced by any of the other changes in this PR. if we can understand exactly why the problem is happening, then we'll have a better idea of whether this is the simplest solution.

Widgets shouldn't even be aware of their parent's background. But in this case, the Box isn't actually the Selection's parent at the native level: both of them are direct children of the RelativeLayout, and the Box appears behind the Selection because of their order of creation. So maybe certain background features are being drawn directly on the native parent (the RelativeLayout), and the intervening Box is covering it up? Experimenting with semi-transparent backgrounds may answer this.

It may also have something to do with [elevation](https://stackoverflow.com/questions/32307245).

_Originally posted by @mhsmith in https://github.com/beeware/toga/issues/2484#issuecomment-2184310959_

proneon267 commented:
Readthedocs test failure is unrelated.

proneon267 commented:
I experimented with semi-transparent backgrounds, to confirm if the arrow and other effects were being drawn on the native parent instead, and @mhsmith was correct:
```
 toga.Box(
            style=Pack(flex=1, background_color=rgba(135, 206, 250, 0.5)),
            children=[
                toga.Selection(
                    items=["Alice", "Bob", "Charlie"],
                )
            ],
        )
```

![Screenshot 2025-01-23 001225](https://github.com/user-attachments/assets/63fe371f-54d0-4826-97d2-be3d110e0896)

The background features are being drawn directly on the native parent(RelativeLayout), as the Box isn't actually the Selection's parent at the native level: both of them are direct children of the RelativeLayout.

To fix this, I have tried the following, but none of them worked:

## Using `setElevation()`:
Note: `setElevation()` takes values in pixels: http://developer.android.com/reference/android/support/v4/view/ViewCompat.html#setElevation(android.view.View,%20float)

- ### Only setting the elevation of the `toga.Selection` widget:
```
selection_widget._impl.native.setElevation(20 * context.getResources().getDisplayMetrics().density)
```
- ### Setting the elevation of both the `toga.Selection` and the `toga.Box` widget:
```
selection_widget._impl.native.setElevation(20 * context.getResources().getDisplayMetrics().density)
box_widget._impl.native.setElevation(1 * context.getResources().getDisplayMetrics().density)
```
- ### Setting the elevation of the `toga.Selection` to positive and the `toga.Box` widget to negative:
```
selection_widget._impl.native.setElevation(20 * context.getResources().getDisplayMetrics().density)
box_widget._impl.native.setElevation(-10 * context.getResources().getDisplayMetrics().density)
```

## Using `bringChildToFront()`:
```
parent = selection_widget._impl.native.getParent()
parent.bringChildToFront(selection_widget._impl.native)
```

## `setTranslationZ()`:
```
selection_widget._impl.native.setTranslationZ(10.0)
```

## Removing and re-adding to the native parent(RelativeLayout):
```
parent = selection_widget.impl.native.getParent()
parent.removeView(selection_widget.impl.native)
parent.addView(selection_widget.impl.native)
```
---

I had also read somewhere that certain features like shadows are only drawn directly on the native parent. As I have also mentioned previously:
https://github.com/beeware/toga/blob/c9c714e3f05d48183a1ace86745812dd6103e3d2/android/src/toga_android/widgets/base.py#L197-L212 :
So, it seems like wrapping the native widget inside another view would allow us to preserve native look and feel, while setting the background color of widgets.

freakboy3742 commented:
I'll need to defer to @mhsmith for suggestions here - my understanding of the inner workings of Android rendering are hazy at best.

mhsmith commented:
Thanks, I'll look at this today or tomorrow.

proneon267 commented:
Thank you :)

mhsmith commented:
> DetailedList widget now correctly unsets the highlight color from an unselected item

This is a good improvement, but I think it's independent of the other issues. So I suggest moving it to a separate PR, and fixing the corresponding problem in Table at the same time.

> setting a custom background_color on the Selection widget, now also sets the background color of the dropdown popup

This probably isn't a good idea, because the two elements will appear against different backgrounds. For example, if you set the background color to transparent, then you get this:

<img src="https://github.com/user-attachments/assets/257ebe66-3ee3-4e57-8f7d-30f8e87ea45a" width="300"/>

I think it's better to treat the popup as a system-provided dialog, like the long-press menu on a TextInput, and not attempt to style it at all.

mhsmith commented:
There's some inconsistency in whether the default background color is transparent or white:

<img src="https://github.com/user-attachments/assets/6c9d7cc9-6a57-4406-83a9-2d61cfe8ea6d" width="300"/>

My inclination is that *everything* should default to transparent, because the Android visual style (at least this version of it) doesn't really emphasize the rectangular borders of widgets.

But I haven't thought through all the implications of this, so if you disagree, let's discuss it before you make any changes.

mhsmith commented:
The ripple effects are supposed to be circular, but they're being clipped by their native parent:

![Screenshot_1738614838](https://github.com/user-attachments/assets/6e9a44aa-d465-454d-ae6d-a205679762ef)

It may be possible to fix this with the "clip" properties (https://stackoverflow.com/questions/30626019).

mhsmith commented:
> My inclination is that everything should default to transparent

This would allow the removal of DEFAULT_BACKGROUND_COLOR and toga_color from colors.py.

It may also allow a significant reduction in the number of set_background overrides in the subclasses, and maybe even allow set_background_color to be implemented in the Widget base class.

mhsmith commented on android/src/toga_android/widgets/base.py:
```diff
@@ -198,3 +186,85 @@ def android_text_align(value):
+# LayerDrawable cannot also be properly cloned. Using `getConstantState()` on the
+# Drawable will produce an erroneous version of the original Drawable.
+#
+# These widgets are also affected by the background color of the parent inside which
+# they are present. Directly, setting background color also destroys the native look
+# and feel of these widgets. Moreover, these effects also serve the purpose of
+# providing visual aid for the action of these widgets.
```
```suggestion
# These widgets also draw some of their background effects on their native parent. 
# But in the Widget base class, the native parent is actually the root Box of the
# layout, and the parent Box is stacked underneath without any native parent/child 
# relationship. So if the parent Box has a background color, it may conceal some of
# the background elements of its children.
```

mhsmith commented on android/src/toga_android/widgets/base.py:
```diff
@@ -198,3 +186,85 @@ def android_text_align(value):
+        # they have been added to a container.
+        self.interface.style.reapply()
+
+    def get_enabled(self):
+        return self.native.isEnabled() and self.native_widget_container.isEnabled()
+
```
Are the `enabled` and `focus` overrides really necessary?

I agree we need something for `hidden`, but could this be done by changing the Widget base class implementation to use `native_toplevel` as discussed above?

proneon267 replied:
> Are the enabled and focus overrides really necessary?
> 
> I agree we need something for hidden, but could this be done by changing the Widget base class implementation to use native_toplevel as discussed above?

They weren't necessary, so I have removed these duplicated methods.

mhsmith commented on android/src/toga_android/widgets/detailedlist.py:
```diff
@@ -96,7 +95,7 @@ def onRefresh(self):
 
 
-class DetailedList(Widget):
+class DetailedList(ContainedWidget):
```
Do DetailedList and ProgressBar need to be ContainedWidget subclasses? I don't think either of them has any special background effects.

proneon267 replied:
I agree, they don't need to be `ContainedWidget`, so I have removed these changes.

mhsmith commented on android/tests_backend/widgets/selection.py:
```diff
@@ -27,10 +27,6 @@ def typeface(self):
     def text_size(self):
         xfail("Can't change the font of Selection on this backend")
 
-    @property
-    def background_color(self):
-        xfail("Can't change the background color of Selection on this backend")
```
This PR enables background colors for quite a few other widgets which didn't support them before. Do their tests also need to be updated?

proneon267 replied:
Yes, the tests should be enabled. However, since `background_color` is not currently correct on all platforms, so we cannot enable the background color tests on this PR, as the tests would fail on the other platforms(as the fix for those platforms is on separate PRs). Since, this would lead to a deadlock situation. Hence, I had created #3015 to keep track of the widgets on which all background color tests are not enabled. Once, these `background_color` fixing PRs are completed, I will enable all background color tests on all platforms.

proneon267 commented:
>>DetailedList widget now correctly unsets the highlight color from an unselected item
>
>This is a good improvement, but I think it's independent of the other issues. So I suggest moving it to a separate PR, and fixing >the corresponding problem in Table at the same time.

I have moved it to #3156.

>>setting a custom background_color on the Selection widget, now also sets the background color of the dropdown popup
>
>This probably isn't a good idea, because the two elements will appear against different backgrounds. For example, if you set the >background color to transparent, then you get this:

I have removed these changes.

> My inclination is that everything should default to transparent (except buttons, obviously), because this version of the Android visual style doesn't really emphasize the rectangular borders of widgets.
> 
> This would allow the removal of DEFAULT_BACKGROUND_COLOR and toga_color from colors.py.
> 
> It may also allow a significant reduction in the number of set_background_color overrides in the subclasses, and maybe even allow the method to be implemented in the Widget base class.
> 
> But I haven't thought through all the implications of this, so if you disagree, let's discuss it before you make any changes.

I totally agree with you, the default background color should be transparent, as is the case on most platforms. Hence, I have made transparent the default background color.

proneon267 commented on android/src/toga_android/container.py:
```diff
@@ -51,13 +51,22 @@ def refreshed(self):
+        if isinstance(widget, ContainedWidget):
+            self.native_content.addView(widget.native_widget_container)
+        else:
+            self.native_content.addView(widget.native)
```
Thanks, this simplified the codebase so much!

proneon267 commented on android/src/toga_android/widgets/base.py:
```diff
@@ -198,3 +186,85 @@ def android_text_align(value):
+# and their native look and feel needs to be preserved.
+class ContainedWidget(Widget):
+    def __init__(self, interface):
+
```
I have refactored it out and have removed all of the duplicated operations.

proneon267 commented:
The only remaining currently is the clipping of the ripple effect:
> The ripple effects are supposed to be circular, but they're being clipped by their native parent:
> 
> ![Screenshot_1738614838](https://private-user-images.githubusercontent.com/166954/409276756-6e9a44aa-d465-454d-ae6d-a205679762ef.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg2ODE4OTMsIm5iZiI6MTczODY4MTU5MywicGF0aCI6Ii8xNjY5NTQvNDA5Mjc2NzU2LTZlOWE0NGFhLWQ0NjUtNDU0ZC1hZTZkLWEyMDU2Nzk3NjJlZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjA0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIwNFQxNTA2MzNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jZGY3MmY4YjdhNDFlMjNkNDYwMzU2MDk3OTQ4ZThhZWU4YjgxNmQzOTk0ODJjMGQ1ZTExNGNlYjQ2YTRkMDY1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9._JvwoFA45heObuyutw-7OEm5PMs35PcyKCJsZfb0-LE)
> 
> It may be possible to fix this with the "clip" properties (https://stackoverflow.com/questions/30626019).

I tried to use the following(on the `native_toplevel` as well as on other parents like container), but they didn't fix the clipping:
```python
self.native_toplevel.setClipChildren(False)
self.native_toplevel.setClipToPadding(False)
self.native_toplevel.setClipToOutline(False)
 ```
However, on further testing, it turns out the problem is due to the incorrect `LayoutParams` for `self.native_toplevel`, hence the ripple effect of `self.native` gets clipped by `self.native_toplevel`. 
<hr>

Currently, we have:
https://github.com/beeware/toga/blob/8959ff1a39d60f433b5b584ad21f28b8e44723ae/android/src/toga_android/container.py#L59-L63
https://github.com/beeware/toga/blob/8959ff1a39d60f433b5b584ad21f28b8e44723ae/android/src/toga_android/widgets/base.py#L218-L230
Which produces the clipped ripple effect:
<details>
  <summary>Example source:(The `Selection` widget's background is intentionally kept white)</summary>

  ```python
"""
My first application
"""

import toga
from toga.style import Pack
from toga.style.pack import COLUMN


class HelloWorld(toga.App):
    def reset_to_native_default_background_color(self, widget, **kwargs):
        for widget in self.widgets:
            del widget.style.background_color
        self.content.refresh()

    def startup(self):
        """Construct and show the Toga application.

        Usually, you would add your application to a main content box.
        We then create a main window (with a name matching the app), and
        show the main window.
        """
        self.content = toga.Box(
            style=Pack(
                background_color="#87CEFA",
                direction=COLUMN,
                flex=1,
            ),
            children=[
                toga.Box(style=Pack(flex=1.5)),
                toga.Button(
                    text="Reset to native default background color",
                    on_press=self.reset_to_native_default_background_color,
                ),
                toga.Box(style=Pack(flex=1.5)),
                toga.ImageView(
                    toga.Image(self.paths.app / "resources/imageview.png"),
                ),
                toga.Box(style=Pack(flex=1.5)),
                toga.Label(text="Label"),
                toga.Box(style=Pack(flex=1.5)),
                toga.ProgressBar(max=100, value=50),
                toga.Box(style=Pack(flex=1.5)),
                toga.Selection(
                    items=["Alice", "Bob", "Charlie"],
                    style=Pack(background_color="#ffffff"),
                ),
                toga.Box(style=Pack(flex=1.5)),
                toga.Slider(min=-5, max=10, value=7),
                toga.Box(style=Pack(flex=1.5)),
                toga.NumberInput(value=8908),
                toga.Box(style=Pack(flex=1.5)),
                toga.MultilineTextInput(
                    value="Some text.\nIt can be multiple lines of text."
                ),
                toga.Box(style=Pack(flex=1.5)),
                toga.TextInput(value="Jane Developer"),
                toga.Box(style=Pack(flex=1.5)),
                toga.PasswordInput(value="Jane"),
                toga.Box(style=Pack(flex=1.5)),
                toga.Table(
                    headings=["Name", "Age"],
                    data=[
                        ("Arthur Dent", 42),
                        ("Ford Prefect", 37),
                        ("Tricia McMillan", 38),
                    ],
                ),
                toga.Box(style=Pack(flex=1.5)),
                toga.Switch(text="Switch"),
                toga.Box(style=Pack(flex=1.5)),
                toga.DetailedList(
                    data=[
                        {
                            "icon": toga.Icon("icons/arthur"),
                            "title": "Arthur Dent",
                            "subtitle": "Where's the tea?",
                        },
                        {
                            "icon": toga.Icon("icons/ford"),
                            "title": "Ford Prefect",
                            "subtitle": "Do you know where my towel is?",
                        },
                        {
                            "icon": toga.Icon("icons/tricia"),
                            "title": "Tricia McMillan",
                            "subtitle": "What planet are you from?",
                        },
                    ],
                ),
                toga.Box(style=Pack(flex=1.5)),
                toga.DateInput(),
                toga.Box(style=Pack(flex=1.5)),
                toga.TimeInput(),
                toga.Box(style=Pack(flex=1.5)),
            ],
        )
        self.main_window = toga.MainWindow(title=self.formal_name)
        self.main_window.content = toga.ScrollContainer(content=self.content)
        self.main_window.show()


def main():
    return HelloWorld()
```
</details>

<img src="https://github.com/user-attachments/assets/6c3cb787-9799-48e9-8ee2-e108a9352b57" width=30%>
<hr>

However, logically the correct code should be:

```diff
diff --git a/android/src/toga_android/container.py b/android/src/toga_android/container.py
index 714e495ac..d46cabfe3 100644
--- a/android/src/toga_android/container.py
+++ b/android/src/toga_android/container.py
@@ -60,4 +60,4 @@ class Container(Scalable):
         lp = RelativeLayout.LayoutParams(width, height)
         lp.topMargin = y
         lp.leftMargin = x
-        widget.native_toplevel.setLayoutParams(lp)
+        widget.native.setLayoutParams(lp)
diff --git a/android/src/toga_android/widgets/base.py b/android/src/toga_android/widgets/base.py
index 2f4774fca..7cd2eb40c 100644
--- a/android/src/toga_android/widgets/base.py
+++ b/android/src/toga_android/widgets/base.py
@@ -222,10 +222,10 @@ class ContainedWidget(Widget):
         self.native_toplevel = RelativeLayout(self._native_activity)
         self.native_toplevel.addView(self.native)
 
-        self.native.setLayoutParams(
+        self.native_toplevel.setLayoutParams(
             RelativeLayout.LayoutParams(
-                RelativeLayout.LayoutParams.MATCH_PARENT,
-                RelativeLayout.LayoutParams.MATCH_PARENT,
+                RelativeLayout.LayoutParams.WRAP_CONTENT,
+                RelativeLayout.LayoutParams.WRAP_CONTENT,
             )
         )
         # Immediately re-apply styles. Some widgets may defer style application until
```

But this produces the wrong result:
<img src="https://github.com/user-attachments/assets/7f4af76c-7fb2-4a2d-84fa-1e2fdefac931" width=30%>

The `native_toplevel` should only enclose upto the height of `self.native` when `native_toplevel` has `WRAP_CONTENT` as layoutparams, but it exceeds beyond the height of `self.native`. Moreover, the bottom height is also not correct, leading to clipping at the bottom part of the ripple effect.

mhsmith commented:
I'm not sure why you think this is related to the LayoutParams. Regardless of whether the parent is sized to the child or vice versa, I expect that if the resulting bounds are the same, then the clipping behavior would be the same too.

> I tried to use the following(on the `native_toplevel` as well as on other parents like container), but they didn't fix the clipping:
> 
> ```python
> self.native_toplevel.setClipChildren(False)
> self.native_toplevel.setClipToPadding(False)
> self.native_toplevel.setClipToOutline(False)
> ```

You're right: I tried these methods, plus `setClipBounds`, on `self.native`, `self.native_toplevel`, and `Container.native_content`, and none of them had any effect. I guess whatever mechanism it's using to draw the ripple on the parent's background, it's being clipped by the parent's bounds regardless of any of these settings.

However, the other comment on the [StackOverflow question](https://stackoverflow.com/questions/30626019) does appear to be true: "The ripple is drawn on the first parent which doesn't have a null background." In the Android API, null and transparent are not the same. If I change `set_background_simple` to call:
 ```py
setBackground(None if color in (None, TRANSPARENT) else ColorDrawable(native_color(color)))
```
Then when the app requests a transparent background, the ripple is drawn correctly on the native parent (the `Container`), and when the background is non-transparent, it's clipped on the `native_toplevel`.

I would be willing to accept this, since we don't recommend setting backgrounds on most widgets anyway. But this still doesn't fix the issue where setting a background color on a Box interferes with the rendering of its children. 

The only solution I can see to that is to actually reproduce the Toga box hierarchy at the native level, rather than making all widgets direct children of the top-level Container. I'm not sure how difficult that would be, but it would be mostly independent of this PR. So I suggest we limit this PR to fixing the *child* background, and make a separate PR later if you want to continue dealing with the *parent* background bug.

proneon267 commented on android/src/toga_android/widgets/base.py:
```diff
@@ -66,6 +64,7 @@ def __init__(self, interface):
             )
         )
 
+        self._default_background_color = Color.TRANSPARENT
```
Thanks, I have inlined it.

proneon267 commented on android/src/toga_android/colors.py:
```diff
@@ -1,7 +1,9 @@
-CACHE = {TRANSPARENT: Color.TRANSPARENT}
+CACHE = {
+    TRANSPARENT: Color.TRANSPARENT,
+}
```
I have reverted it.

proneon267 commented on android/tests_backend/widgets/base.py:
```diff
@@ -27,15 +26,18 @@ def __init__(self, widget):
-        assert self.native.getParent() is container._impl.container.native_content
+        assert (
+            self.native_toplevel.getParent() is container._impl.container.native_content
+        )
```
I have added a new check.

proneon267 commented on android/tests_backend/widgets/base.py:
```diff
@@ -95,14 +97,14 @@ def assert_layout(self, size, position):
-            # minimum API level.
+            # The following complex Drawables all apply color filters to
+            # their children, but they don't implement getColorFilter, at
+            # least not in our current minimum API level.
```
Thanks, I have reverted these changes.

proneon267 commented on android/tests_backend/widgets/properties.py:
```diff
@@ -18,7 +18,7 @@ def toga_color(color_int):
             Color.green(color_int),
             Color.blue(color_int),
-            Color.alpha(color_int) / 255,
+            round(Color.alpha(color_int) / 255, 2),
```
Thanks, I have reverted it.

proneon267 commented on android/src/toga_android/widgets/base.py:
```diff
@@ -128,33 +131,20 @@ def set_font(self, font):
     # using one of the utility functions below.
     def set_background_color(self, color):
-        pass
+        self.set_background_simple(color)
```
Thanks, I have changed it.

proneon267 commented:
> Then when the app requests a transparent background, the ripple is drawn correctly on the native parent (the Container), and when the background is non-transparent, it's clipped on the native_toplevel. I would be willing to accept this, since we don't recommend setting backgrounds on most widgets anyway.
> 
> But this still doesn't fix the issue where setting a background color on a Box interferes with the rendering of its children. The only solution I can see to that is to actually reproduce the Toga box hierarchy at the native level, rather than making all widgets direct children of the top-level Container. I'm not sure how difficult that would be, but it would be mostly independent of this PR. So I suggest we limit this PR to fixing the child backgrounds, and make a separate PR later if you want to continue dealing with the parent background bug.

I agree with you, the actual solution to fixing the issue would be to recreate the widget hierarchy at the native level. As you have noted, I will do it in a separate PR.

mhsmith commented on android/src/toga_android/widgets/base.py:
```diff
@@ -126,21 +125,19 @@ def set_font(self, font):
 
     def set_background_filter(self, color):
+        # Although setBackgroundColor is defined in the View base class, we can't
+        # use it as a default implementation on some widgets(e.g. Button), because
+        # it often overwrites other aspects of the widget's appearance. For example,
+        # when setBackgroundColor is used on a Button, it changes the background behind
+        # the button rather than the button's actual color.
```
```suggestion
        # use it as a default implementation on some widgets (e.g. Button), because
        # it often overwrites other aspects of the widget's appearance. For example,
        # when setBackgroundColor is used on a Button, it makes the button appear
        # as a solid rectangle with no bevels or animations.
```

mhsmith commented on android/src/toga_android/widgets/base.py:
```diff
@@ -194,3 +181,44 @@ def android_text_align(value):
     }[value]
+
+
+# Most of the Android Widget have different effects applied them which provide
+# the native look and feel of Android. These widgets' background consists of
+# Drawables like ColorDrawable, InsetDrawable and other animation Effect Drawables
+# like RippleDrawable. Often when such Effect Drawables are used, they are stacked
```
```suggestion
# The look and feel of Android widgets is sometimes implemented using background
# Drawables like ColorDrawable, InsetDrawable and other animation effect Drawables
# like RippleDrawable. Often when such effect Drawables are used, they are stacked
```

mhsmith commented on android/src/toga_android/widgets/base.py:
```diff
@@ -194,3 +181,44 @@ def android_text_align(value):
+# LayerDrawable cannot also be properly cloned. Using `getConstantState()` on the
+# Drawable will produce an erroneous version of the original Drawable.
+#
+# These widgets also draw some of their background effects on their native parent.
+# But in the Widget base class, the native parent is actually the root Box of the
+# layout, and the parent Box is stacked under its children without any native
+# parent/child relationship. So if the parent Box has a background color, it may
+# conceal some of the background elements of its children.
+#
```
```suggestion
```

mhsmith commented:
Thanks very much; I've created #3235 to record the remaining issue.
